### PR TITLE
v0.3.0 update

### LINKED ISSUE #19: Charon update
We should update the mobile wrapper's go-perun to Charon.

This implies several changes/additions
- Expose `OnNewChannel` callback registration. Document that starting of watcher should happen here, if wanted.
- Change to and expose new common handler routine `Client.Handle`.
- Add an optional `dbpath` section to the configuration. If set, use it to enable leveldb persistence.
- Expose `Client.Reconnect(*Context)` method. Note that the `OnNewChannel` callback has to be set _before_ restoring of channels from persistence happens, or persisted channels go unnoticed and are only pollable through `Client.Channel`.
- Apply all changes to the Java example.

### PR DESCRIPTION
**Pull request by ggwpez**

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -77,6 +77,23 @@ func NewClient(ctx *Context, cfg *Config, w *Wallet) (*Client, error) {
 	return &Client{cfg: cfg, ethClient: ethClient, client: c, w: w.w, onChain: acc, dialer: dialer}, nil
 }
 
+// Handle is the handler routine for channel proposals and channel updates. It
+// must only be started at most once by the user.
+// Incoming proposals and updates are forwarded to the passed handlers.
```
Comment (ggwpez) [inline client.go:97]: Will add links

**Comment by ggwpez** [comment]
Please focus the review on the bindings since we add the `Channel.Peers()` function and that will change the example. @sebastianst

File: client.go [hunk 2]
```diff
@@ -22,21 +23,32 @@ import (
...
-	dialer *net.Dialer
-}
+		dialer *net.Dialer
+	}
+
+	// ChannelAcceptor used to wrap a `func(*PaymentChannel)`
+	// function pointer for the `OnNewChannel` callback.
+	ChannelAcceptor interface {
+		Accept(*PaymentChannel)
+	}
```
Comment (sebastianst) [inline client.go]: The naming is very confusing. This is not an `Acceptor` but only a callback/event consumer. There's nothing to accept here, the channel already got accepted and created. Please change this to something like
```suggestion
	// NewChannelCallback wraps a `func(*PaymentChannel)`
	// function pointer for the `Client.OnNewChannel` callback.
	NewChannelCallback interface {
		OnNew(*PaymentChannel)
	}
```
Also change the naming everywhere.

File: client.go [hunk 3]
```diff
@@ -77,6 +89,40 @@ func NewClient(ctx *Context, cfg *Config, w *Wallet) (*Client, error) {
...
+// or restored. Only one such handler can be set at a time, and repeated calls
+// to this function will overwrite the currently existing handler. This
+// function may be safely called at any time.
+// Start the watcher routine here, if needed.
+// ref https://pkg.go.dev/perun.network/go-perun/client?tab=doc#Client.OnNewChannel
+func (c *Client) OnNewChannel(acc ChannelAcceptor) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	c.acceptor = acc
```
Comment (sebastianst) [inline client.go]: Remove `c.acceptor` field from type `Client`. Always keep as few state as possible. It is only used here locally to set the `OnNewChannel` callback.

File: client.go [hunk 4]
```diff
@@ -77,6 +89,40 @@ func NewClient(ctx *Context, cfg *Config, w *Wallet) (*Client, error) {
...
+	c.client.OnNewChannel(func(ch *client.Channel) {
+		c.mtx.RLock()
+		defer c.mtx.RUnlock()
+
+		c.acceptor.Accept(&PaymentChannel{ch})
```
Comment (sebastianst) [inline client.go]:
```suggestion
		callback.Accept(&PaymentChannel{ch})
```

File: client.go [hunk 5]
```diff
@@ -77,6 +89,40 @@ func NewClient(ctx *Context, cfg *Config, w *Wallet) (*Client, error) {
...
+	c.client.Handle(&proposalHandler{c: c, h: ph}, &updateHandler{h: uh})
+}
+
+// OnNewChannel sets a handler to be called whenever a new channel is created
+// or restored. Only one such handler can be set at a time, and repeated calls
+// to this function will overwrite the currently existing handler. This
+// function may be safely called at any time.
+// Start the watcher routine here, if needed.
+// ref https://pkg.go.dev/perun.network/go-perun/client?tab=doc#Client.OnNewChannel
+func (c *Client) OnNewChannel(acc ChannelAcceptor) {
```
Comment (sebastianst) [inline client.go]:
```suggestion
func (c *Client) OnNewChannel(callback NewChannelCallback) {
```

File: client.go [hunk 6]
```diff
@@ -22,21 +23,32 @@ import (
...
+	Client struct {
+		cfg *Config
 
-	ethClient *ethclient.Client
-	client    *client.Client
+		ethClient *ethclient.Client
+		client    *client.Client
+		// mtx protects the acceptor
+		mtx      sync.RWMutex
+		acceptor ChannelAcceptor
```
Comment (sebastianst) [inline client.go]: Remove `c.acceptor` field from type `Client`. Always keep as few state as possible.
```suggestion
```

**Review by sebastianst — CHANGES_REQUESTED** [review]
Example looks good. Please rename the Acceptor to a callback, as mentioned in the review comments.

**Comment by ggwpez** [comment]
I have to change the example to the new interface naming… We really need a pipeline :laughing:

**Review by sebastianst — APPROVED** [review]

**Pull request merged**